File: packages/google_mlkit_digital_ink_recognition/ios/Classes/GoogleMlKitDigitalInkRecognitionPlugin.m
#import "GoogleMlKitDigitalInkRecognitionPlugin.h"
#import <MLKitCommon/MLKitCommon.h>
#import <MLKitDigitalInkRecognition/MLKitDigitalInkRecognition.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#define channelName @"google_mlkit_digital_ink_recognizer"
#define startDigitalInkRecognizer @"vision#startDigitalInkRecognizer"
#define closeDigitalInkRecognizer @"vision#closeDigitalInkRecognizer"
#define manageInkModels @"vision#manageInkModels"

@implementation GoogleMlKitDigitalInkRecognitionPlugin {
    NSMutableDictionary *instances;
    GenericModelManager *genericModelManager;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitDigitalInkRecognitionPlugin* instance = [[GoogleMlKitDigitalInkRecognitionPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startDigitalInkRecognizer]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:manageInkModels]) {
        [self manageModel:call result:result];
    } else if ([call.method isEqualToString:closeDigitalInkRecognizer]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *modelTag = call.arguments[@"model"];
    
    MLKDigitalInkRecognitionModelIdentifier *identifier = [MLKDigitalInkRecognitionModelIdentifier modelIdentifierForLanguageTag:modelTag];
    MLKDigitalInkRecognitionModel *model = [[MLKDigitalInkRecognitionModel alloc] initWithModelIdentifier:identifier];
    MLKModelManager *modelManager = [MLKModelManager modelManager];
    
    BOOL isModelDownloaded = [modelManager isModelDownloaded:model];
    
    if (!isModelDownloaded) {
        FlutterError *error = [FlutterError errorWithCode:@"Error Model has not been downloaded yet"
                                                  message:@"Model has not been downloaded yet"
                                                  details:@"Model has not been downloaded yet"];
        result(error);
        return;
    }
    
    NSString *uid = call.arguments[@"id"];
    MLKDigitalInkRecognizer *recognizer = [instances objectForKey:uid];
    if (recognizer == NULL) {
        MLKDigitalInkRecognizerOptions *options = [[MLKDigitalInkRecognizerOptions alloc] initWithModel:model];
        recognizer = [MLKDigitalInkRecognizer digitalInkRecognizerWithOptions:options];
        instances[uid] = recognizer;
    }
    
    NSMutableArray *strokes = [NSMutableArray array];
    NSArray *strokeList = call.arguments[@"ink"][@"strokes"];
    for (NSDictionary *strokeMap in strokeList) {
        NSMutableArray *stroke = [NSMutableArray array];
        NSArray *pointsList = strokeMap[@"points"];
        for (NSDictionary *pointMap in pointsList) {
            NSNumber *x = pointMap[@"x"];
            NSNumber *y = pointMap[@"y"];
            NSNumber *t = pointMap[@"t"];
            MLKStrokePoint *strokePoint = [[MLKStrokePoint alloc] initWithX:x.floatValue y:y.floatValue t:t.longValue];
            [stroke addObject:strokePoint];
        }
        [strokes addObject:[[MLKStroke alloc] initWithPoints:stroke]];
    }
    MLKInk *ink = [[MLKInk alloc] initWithStrokes:strokes];
    
    MLKDigitalInkRecognitionContext *context;
    NSDictionary *contextMap = call.arguments[@"context"];
    if ([contextMap isKindOfClass: [NSDictionary class]]) {
        NSString *preContext = contextMap[@"preContext"];
        if ([preContext isKindOfClass: [NSNull class]]) {
            preContext = @"";
        }
        MLKWritingArea *writingArea;
        NSDictionary *writingAreaMap = contextMap[@"writingArea"];
        if ([writingAreaMap isKindOfClass: [NSDictionary class]]) {
            NSNumber *width = writingAreaMap[@"width"];
            NSNumber *height = writingAreaMap[@"height"];
            writingArea = [[MLKWritingArea alloc] initWithWidth:width.floatValue height:height.floatValue];
        }
        context = [[MLKDigitalInkRecognitionContext alloc] initWithPreContext:preContext writingArea:writingArea];
    }
    
    if (context != NULL) {
        [recognizer recognizeInk:ink
                         context:context
                      completion:^(MLKDigitalInkRecognitionResult * _Nullable recognitionResult,
                                   NSError * _Nullable error) {
            [self process:recognitionResult error:error result:result];
        }];
    } else {
        [recognizer recognizeInk:ink
                      completion:^(MLKDigitalInkRecognitionResult * _Nullable recognitionResult,
                                   NSError * _Nullable error) {
            [self process:recognitionResult error:error result:result];
        }];
    }
}

- (void )process:(MLKDigitalInkRecognitionResult *)recognitionResult
           error:(NSError *)error
          result:(FlutterResult)result {
    if (error) {
        result(getFlutterError(error));
        return;
    } else if (!recognitionResult) {
        result(NULL);
        return;
    }
    NSMutableArray *candidates = [NSMutableArray new];
    for(MLKDigitalInkRecognitionCandidate *candidate in recognitionResult.candidates) {
        NSDictionary *dictionary = @{@"text": candidate.text,
                                     @"score": @(candidate.score.doubleValue)};
        [candidates addObject:dictionary];
    }
    result(candidates);
}

- (void)manageModel:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *modelTag = call.arguments[@"model"];
    MLKDigitalInkRecognitionModelIdentifier *identifier = [MLKDigitalInkRecognitionModelIdentifier modelIdentifierForLanguageTag:modelTag];
    MLKDigitalInkRecognitionModel *model = [[MLKDigitalInkRecognitionModel alloc] initWithModelIdentifier:identifier];
    genericModelManager = [[GenericModelManager alloc] init];
    [genericModelManager manageModel:model call:call result:result];
}

@end

File: packages/google_mlkit_digital_ink_recognition/ios/Classes/GoogleMlKitDigitalInkRecognitionPlugin.swift
import Flutter
import MLKitCommon
import MLKitDigitalInkRecognition
import google_mlkit_commons

@objc
public class GoogleMlKitDigitalInkRecognitionPlugin: NSObject, FlutterPlugin {
  private var instances: [String: DigitalInkRecognizer] = [:]
  private var genericModelManager: GenericModelManager?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_digital_ink_recognizer",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitDigitalInkRecognitionPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "vision#startDigitalInkRecognizer":
      handleDetection(call: call, result: result)
    case "vision#manageInkModels":
      manageModel(call: call, result: result)
    case "vision#closeDigitalInkRecognizer":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let modelTag = args["model"] as? String,
          let uid = args["id"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
      return
    }

    guard let identifier = DigitalInkRecognitionModelIdentifier(forLanguageTag: modelTag) else {
      result(FlutterError(code: "invalid_model", message: "Invalid language tag: \(modelTag)", details: nil))
      return
    }
    let model = DigitalInkRecognitionModel(modelIdentifier: identifier)
    guard ModelManager.modelManager().isModelDownloaded(model) else {
      result(FlutterError(
        code: "Error Model has not been downloaded yet",
        message: "Model has not been downloaded yet",
        details: "Model has not been downloaded yet"
      ))
      return
    }

    let recognizer: DigitalInkRecognizer
    if let existing = instances[uid] {
      recognizer = existing
    } else {
      let options = DigitalInkRecognizerOptions(model: model)
      recognizer = DigitalInkRecognizer.digitalInkRecognizer(options: options)
      instances[uid] = recognizer
    }

    guard let strokeList = args["ink"] as? [String: Any],
          let strokesData = strokeList["strokes"] as? [[String: Any]] else {
      result(FlutterError(code: "invalid_args", message: "Missing ink data", details: nil))
      return
    }
    let strokes = strokesData.map { strokeMap -> Stroke in
      guard let pointsList = strokeMap["points"] as? [[String: Any]] else {
        return Stroke(points: [])
      }
      let points = pointsList.map { pointMap -> StrokePoint in
        let coordX = (pointMap["x"] as? NSNumber)?.floatValue ?? 0
        let coordY = (pointMap["y"] as? NSNumber)?.floatValue ?? 0
        let timeMs = (pointMap["t"] as? NSNumber)?.intValue ?? 0
        return StrokePoint(x: coordX, y: coordY, t: timeMs)
      }
      return Stroke(points: points)
    }
    let ink = Ink(strokes: strokes)

    let contextMap = args["context"] as? [String: Any]
    var context: DigitalInkRecognitionContext?
    if let ctx = contextMap {
      let preContext = ctx["preContext"] as? String ?? ""
      var writingArea: WritingArea?
      if let writingAreaMap = ctx["writingArea"] as? [String: Any],
         let width = writingAreaMap["width"] as? NSNumber,
         let height = writingAreaMap["height"] as? NSNumber {
        writingArea = WritingArea(width: width.floatValue, height: height.floatValue)
      }
      context = DigitalInkRecognitionContext(preContext: preContext, writingArea: writingArea)
    }

    func process(recognitionResult: DigitalInkRecognitionResult?, error: Error?) {
      if let error = error {
        let nsError = error as NSError
        result(FlutterError(
          code: "Error \(nsError.code)",
          message: nsError.domain,
          details: nsError.localizedDescription
        ))
        return
      }
      guard let recognitionResult = recognitionResult else {
        result(nil)
        return
      }
      let candidates = recognitionResult.candidates.map { candidate in
        [
          "text": candidate.text,
          "score": candidate.score?.doubleValue ?? 0
        ] as [String: Any]
      }
      result(candidates)
    }

    if let ctx = context {
      recognizer.recognize(ink: ink, context: ctx) { recognitionResult, error in
        process(recognitionResult: recognitionResult, error: error)
      }
    } else {
      recognizer.recognize(ink: ink) { recognitionResult, error in
        process(recognitionResult: recognitionResult, error: error)
      }
    }
  }

  private func manageModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let modelTag = args["model"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing model argument", details: nil))
      return
    }
    guard let identifier = DigitalInkRecognitionModelIdentifier(forLanguageTag: modelTag) else {
      result(FlutterError(code: "invalid_model", message: "Invalid language tag: \(modelTag)", details: nil))
      return
    }
    let model = DigitalInkRecognitionModel(modelIdentifier: identifier)
    let manager = GenericModelManager()
    genericModelManager = manager
    manager.manage(model: model, call: call, result: result)
  }
}
